// @speechly/react-client/dist/context.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpeechProvider = exports.SpeechContext = void 0;
var react_1 = __importDefault(require("react"));
var browser_client_1 = require("@speechly/browser-client");
/**
 * A React context that holds the state of Speechly SLU API client.
 * @public
 */
exports.SpeechContext = react_1.default.createContext({
    connect: function () { return __awaiter(void 0, void 0, void 0, function () { return __generator(this, function (_a) {
        return [2 /*return*/, Promise.resolve()];
    }); }); },
    attachMicrophone: function () { return __awaiter(void 0, void 0, void 0, function () { return __generator(this, function (_a) {
        return [2 /*return*/, Promise.resolve()];
    }); }); },
    start: function () { return __awaiter(void 0, void 0, void 0, function () { return __generator(this, function (_a) {
        return [2 /*return*/, Promise.resolve('Unknown contextId')];
    }); }); },
    stop: function () { return __awaiter(void 0, void 0, void 0, function () { return __generator(this, function (_a) {
        return [2 /*return*/, Promise.resolve('Unknown contextId')];
    }); }); },
    clientState: browser_client_1.DecoderState.Disconnected,
    microphoneState: browser_client_1.AudioSourceState.Stopped,
    listening: false,
});
/**
 * The provider for SpeechContext.
 *
 * Make sure you have only one SpeechProvider in your application,
 * because otherwise the audio will be mixed up and unusable.
 *
 * It is possible to switch the props on the fly, which will make provider stop current client if it's running
 * and start a new one.
 * @public
 */
var SpeechProvider = /** @class */ (function (_super) {
    __extends(SpeechProvider, _super);
    function SpeechProvider(props) {
        var _this = _super.call(this, props) || this;
        // Client instance stored for immediate access - React 18 will mount, unmount and mount in rapid succession
        _this.client = undefined;
        // init at mount mount time for to play it nice with SSR
        _this.componentDidMount = function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.createClient(this.props);
                return [2 /*return*/];
            });
        }); };
        _this.connect = function () { return __awaiter(_this, void 0, void 0, function () {
            var client;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        client = this.state.client;
                        if (client == null) {
                            throw Error('No Speechly client (are you calling connect in non-browser environment)');
                        }
                        return [4 /*yield*/, client.initialize()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); };
        _this.attachMicrophone = function () { return __awaiter(_this, void 0, void 0, function () {
            var client, microphone;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        client = this.state.client;
                        if (client == null) {
                            throw Error('No Speechly client (are you calling connect in non-browser environment)');
                        }
                        microphone = new browser_client_1.BrowserMicrophone();
                        microphone.onStateChange(function (state) {
                            _this.setState({
                                microphoneState: state,
                            });
                        });
                        return [4 /*yield*/, microphone.initialize()];
                    case 1:
                        _a.sent();
                        if (!microphone.mediaStream) return [3 /*break*/, 3];
                        return [4 /*yield*/, client.attach(microphone.mediaStream)];
                    case 2:
                        _a.sent();
                        return [3 /*break*/, 4];
                    case 3: throw Error('Microphone contains no MediaStream to attach');
                    case 4:
                        this.setState({
                            microphone: microphone,
                        });
                        return [2 /*return*/];
                }
            });
        }); };
        _this.start = function () { return __awaiter(_this, void 0, void 0, function () {
            var _a, client, appId;
            return __generator(this, function (_b) {
                _a = this.state, client = _a.client, appId = _a.appId;
                this.setState({ listening: true });
                if (client == null) {
                    throw Error('No Speechly client (are you calling start in non-browser environment)');
                }
                if (appId !== undefined) {
                    return [2 /*return*/, client.start({ appId: appId })];
                }
                return [2 /*return*/, client.start()];
            });
        }); };
        _this.stop = function () { return __awaiter(_this, void 0, void 0, function () {
            var client;
            return __generator(this, function (_a) {
                client = this.state.client;
                this.setState({ listening: false });
                if (client == null) {
                    throw Error('No Speechly client (are you calling stop in non-browser environment)');
                }
                return [2 /*return*/, client.stop()];
            });
        }); };
        _this.createClient = function (clientOptions) {
            var _a;
            // Postpone connect
            var effectiveOpts = __assign(__assign({}, clientOptions), { connect: false });
            _this.client = new browser_client_1.BrowserClient(effectiveOpts);
            _this.client.onStateChange(_this.onClientStateChange);
            _this.client.onSegmentChange(_this.onSegmentChange);
            _this.client.onTranscript(_this.onTranscript);
            _this.client.onEntity(_this.onEntity);
            _this.client.onIntent(_this.onIntent);
            if (!((_a = clientOptions.disableTentative) !== null && _a !== void 0 ? _a : false)) {
                _this.client.onTentativeIntent(_this.onTentativeIntent);
                _this.client.onTentativeTranscript(_this.onTentativeTranscript);
                _this.client.onTentativeEntities(_this.onTentativeEntities);
            }
            // Connect now to pre-warm backend if not explicitely told not to
            if (clientOptions.connect !== false) {
                // eslint-disable-next-line @typescript-eslint/no-floating-promises
                _this.client.initialize();
            }
            _this.setState({
                client: _this.client,
            });
        };
        _this.onClientStateChange = function (clientState) {
            if (clientState <= browser_client_1.DecoderState.Disconnected) {
                _this.setState({ listening: false });
            }
            _this.setState({ clientState: clientState });
        };
        _this.onSegmentChange = function (segment) {
            if (!segment.isFinal) {
                _this.setState({ segment: segment });
                return;
            }
            // Reset individual values when a segment is finalised.
            _this.setState({
                segment: segment,
                tentativeTranscript: undefined,
                transcript: undefined,
                tentativeEntities: undefined,
                entity: undefined,
                tentativeIntent: undefined,
                intent: undefined,
            });
        };
        _this.onTentativeTranscript = function (contextId, segmentId, words, text) {
            _this.setState({
                tentativeTranscript: {
                    contextId: contextId,
                    segmentId: segmentId,
                    text: text,
                    words: words,
                },
            });
        };
        _this.onTranscript = function (contextId, segmentId, word) {
            _this.setState({
                transcript: {
                    contextId: contextId,
                    segmentId: segmentId,
                    word: word,
                },
            });
        };
        _this.onTentativeEntities = function (contextId, segmentId, entities) {
            _this.setState({
                tentativeEntities: {
                    contextId: contextId,
                    segmentId: segmentId,
                    entities: entities,
                },
            });
        };
        _this.onEntity = function (contextId, segmentId, entity) {
            _this.setState({
                entity: {
                    contextId: contextId,
                    segmentId: segmentId,
                    entity: entity,
                },
            });
        };
        _this.onTentativeIntent = function (contextId, segmentId, intent) {
            _this.setState({
                tentativeIntent: {
                    contextId: contextId,
                    segmentId: segmentId,
                    intent: intent,
                },
            });
        };
        _this.onIntent = function (contextId, segmentId, intent) {
            _this.setState({
                intent: {
                    contextId: contextId,
                    segmentId: segmentId,
                    intent: intent,
                },
            });
        };
        _this.state = {
            client: undefined,
            microphone: undefined,
            listening: false,
            clientState: browser_client_1.DecoderState.Disconnected,
            microphoneState: browser_client_1.AudioSourceState.Stopped,
            appId: props.appId,
        };
        return _this;
    }
    SpeechProvider.prototype.render = function () {
        return (react_1.default.createElement(exports.SpeechContext.Provider, { value: {
                connect: this.connect,
                attachMicrophone: this.attachMicrophone,
                start: this.start,
                stop: this.stop,
                appId: this.state.appId,
                listening: this.state.listening,
                clientState: this.state.clientState,
                microphoneState: this.state.microphoneState,
                segment: this.state.segment,
                tentativeTranscript: this.state.tentativeTranscript,
                transcript: this.state.transcript,
                tentativeEntities: this.state.tentativeEntities,
                entity: this.state.entity,
                tentativeIntent: this.state.tentativeIntent,
                intent: this.state.intent,
                client: this.state.client,
                microphone: this.state.microphone,
            } }, this.props.children));
    };
    SpeechProvider.prototype.componentDidUpdate = function (prevProps) {
        return __awaiter(this, void 0, void 0, function () {
            var props, _a, client, microphone, e_1, e_2;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        props = this.props;
                        // We cannot compare microphone / api client / storage implementations,
                        // so changes in those will be ignored.
                        // A better approach for those would be to use separate contexts.
                        if (props.appId === prevProps.appId &&
                            props.sampleRate === prevProps.sampleRate &&
                            props.debug === prevProps.debug &&
                            props.apiUrl === prevProps.apiUrl) {
                            return [2 /*return*/];
                        }
                        _a = this.state, client = _a.client, microphone = _a.microphone;
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, (client === null || client === void 0 ? void 0 : client.close())];
                    case 2:
                        _b.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        e_1 = _b.sent();
                        console.error('Error closing Speechly client:', e_1);
                        return [3 /*break*/, 4];
                    case 4:
                        _b.trys.push([4, 6, , 7]);
                        return [4 /*yield*/, (microphone === null || microphone === void 0 ? void 0 : microphone.close())];
                    case 5:
                        _b.sent();
                        return [3 /*break*/, 7];
                    case 6:
                        e_2 = _b.sent();
                        console.error('Error closing microphone:', e_2);
                        return [3 /*break*/, 7];
                    case 7:
                        this.createClient(props);
                        return [2 /*return*/];
                }
            });
        });
    };
    SpeechProvider.prototype.componentWillUnmount = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var _b;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _c.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, ((_a = this.client) === null || _a === void 0 ? void 0 : _a.close())];
                    case 1:
                        _c.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        _b = _c.sent();
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    return SpeechProvider;
}(react_1.default.Component));
exports.SpeechProvider = SpeechProvider;
